src/main/java/io/anserini/rerank/AxiomReranker.java

```diff
@@ -0,0 +1,306 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.Terms;
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.queryparser.flexible.standard.StandardQueryParser;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TopDocs;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+import java.util.Random;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+public class AxiomReranker implements Reranker {
+    private static final Logger LOG = LogManager.getLogger(AxiomReranker.class);
+
+    private final Analyzer analyzer;
+    private final String field;
+
+    private int R = 30; // factor that used in extracting random documents, we will extract R*M randomly select documents
+    private int M = 20; // number of top documents in initial results
+    private int L = 1000; // top similar terms
+    private int K = 20; // number of expansion terms
+    private float beta; // scaling parameter
+
+    public AxiomReranker(Analyzer analyzer, String field, String stoplist, Boolean fromResource, float beta) {
+        this.analyzer = analyzer;
+        this.field = field;
+        this.beta = beta;
+    }
+
+    @Override
+    public ScoredDocuments rerank(ScoredDocuments docs, RerankerContext context) {
+        Preconditions.checkState(docs.documents.length == docs.scores.length);
+
+        try {
+            // Select R*M docs from the original ranking list as the reranking pool
+            Set<Integer> usedDocs = selectDocs(docs, context);
+            // Extract an inverted list from the reranking pool
+            Map<String, Set<Integer>> termInvertedList = extractTerms(usedDocs, context, null);
+            // Calculate all the terms in the reranking pool and pick top K of them
+            Map<String, Double> expandedTermScores = calTermScore(termInvertedList, context);
+            StringBuilder builder = new StringBuilder();
+            for (Map.Entry<String, Double> termScore : expandedTermScores.entrySet()) {
+                String term = termScore.getKey();
+                double score = termScore.getValue();
+                builder.append(term + "^" + score + " ");
+            }
+            String queryText = builder.toString().trim();
+
+            if (queryText.isEmpty()) {
+                LOG.info("[Empty Expanded Query]: " + context.getQueryTokens());
+                queryText = context.getQueryText();
+            }
+
+            StandardQueryParser p = new StandardQueryParser();
```

Comment:
The expansion query is explicitly constructed by us - that means we do know there is no characters that can break the query parser.